src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -119,6 +121,90 @@ void TieredHNSWIndex<DataType, DistType>::executeInsertJobWrapper(void *job) {
         ->executeInsertJob(insert_job);
 }
 
+template <typename DataType, typename DistType>
+HNSWIndex<DataType, DistType> *TieredHNSWIndex<DataType, DistType>::getHNSWIndex() {
+    return reinterpret_cast<HNSWIndex<DataType, DistType> *>(this->index);
+}
+
+/******************** Job's callbacks **********************************/
+template <typename DataType, typename DistType>
+void TieredHNSWIndex<DataType, DistType>::executeInsertJob(HNSWInsertJob *job) {
+    HNSWIndex<DataType, DistType> *hnsw_index = this->getHNSWIndex();
+    this->flatIndexGuard.lock_shared();
+    if (job->label == HNSW_INVALID_LABEL) {
+        // Job has been invalidated in the meantime.
+        this->flatIndexGuard.unlock_shared();
+        goto finish;
+    }
+    // Acquire the index data lock, so we know what is the exact index size at this time.
+    this->mainIndexGuard.lock_shared();
```

Comment:
Because of the `addVector` call to HNSW that should be protected with a read lock. Also, to avoid deadlocks, we always acquire the main index lock *before* we take internal HNSW locks.